### PR TITLE
feat: add resource pool state streaming API

### PR DESCRIPTION
**Pull request by armandmcqueen**
## Description

Add a streaming API that sends updates whenever the resource pools change. Resource pools are mostly static so the only updates that can happen are agents joining/leaving and tasks being started/completed. 

This API is separate from the ResourcePool REST API because the REST API is very stable, while the streaming API could change in the future and this approach will make it easier to change one without impacting the other. 

## Commentary (optional)

I'm not certain this is working correctly yet.

## Checklist

- [ ] User-facing API changes need the "User-facing API Change" label.
- [ ] Release notes should be added as a separate file under `docs/release-notes/`.
See [Release Note](../docs/release-notes/README.md) for details.

[DET-1234]: https://determinedai.atlassian.net/browse/DET-1234

### CODE REVIEW

**Comment by hamidzr** [comment]
let's talk about this.
would it work if this was https://github.com/determined-ai/determined/pull/1724 + a field mask to show which fields are set/changed?

**Comment by armandmcqueen** [comment]
@hamidzr This PR is now ready to review

**Review by hamidzr — COMMENTED** [review]
I had forgot to submit. I have yet to run the branch and I'm working on it now ( I was running into issues with db migration)

File: master/internal/api_resourcepool.go [hunk 1]
```diff
@@ -28,3 +36,150 @@ func (a *apiServer) GetResourcePools(
...
+				CpuContainersRunning: int32(summary.CPUContainersRunning),
+			}
+			currentStates = append(currentStates, &state)
+		}
+
+		// Convert full state into diff to send to client
+		updatesToSend := calculateUpdates(a.lastStates, currentStates)
+		a.lastStates = currentStates
+
+		err := a.resp.Send(&apiv1.ResourcePoolStatesResponse{
```
Comment (hamidzr) [inline master/internal/api_resourcepool.go:134]: suggestion: wdy about only sending updates when there is at least one?

**Review by hamidzr — COMMENTED** [review]
overall looks good to me and worked nicely after I ran the branch. I added some small comments. one thing I think could be good to discuss is to avoid sending empty updates.

File: master/internal/api_resourcepool.go (hunk 1 above)
Comment (armandmcqueen) [reply master/internal/api_resourcepool.go:134]: Yeah that's a good idea

**Comment by armandmcqueen** [comment]
Closing until this get addressed during tech debt week

**Pull request closed without merging**